Separate and Update Examples

waciumawanjohi opened the pull request:
---
Name: Separate and Update Examples
About: Creates separate example for Runnable, example for simple supply chain, example for supply chain with runnable.

---
<!--
Thanks for submitting a pull request to Cartographer!

Also check the [PR guidelines] if you haven't already!
-->

[PR requirements]: https://github.com/vmware-tanzu/cartographer/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines

## Changes proposed by this PR

<!--
Add the story that is resolved or updated
`closes`/`fixes` or `updates` are valid here 
-->
Resolves #402

<!--
Summarize your changes. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->


## Release Note

- Separate basic supply chain example from supply chain with testing example.
- Provide example of runnable.
<!--
Your PR title will be directly included in the release notes when it ships in
the next release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Add CreatorName field in Workload spec

Example notes:

* Operators can label stamped objects with the Workload's creator
* Creators can be contacted when interdepartmental issues arise 

If there are no additional notes necessary you may remove this entire section.
-->
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [x] Removed non-atomic or `wip` commits
- [x] Filled in the [Release Note](#Release-Note) section above 
- [x] Modified the docs to match changes <!-- TBD: reference doc editing guidance -->

emmjohnson commented on examples/README.md:
```diff
@@ -1,6 +1,24 @@
+tekton task that tests source code and demonstrates how to create the taskruns with
+objects that don't require triggers or events.
+
+## Supply Chain Examples
```
Move Supply Chain Examples above Runnable Example?

emmjohnson replied:
After going thru each one, it makes most sense to have basic supply chain, runnable, testing supply chain?

emmjohnson commented on examples/README.md:
```diff
@@ -1,6 +1,24 @@
+
+The [runnable-tekton](runnable-tekton/README.md) example demonstrates how Cartographer's
+`Runnable` wraps immutable objects to provide updatable behavior. The example uses a 
+tekton task that tests source code and demonstrates how to create the taskruns with
```
Tekton

emmjohnson replied:
Link to TaskRuns - https://tekton.dev/docs/pipelines/taskruns/

emmjohnson commented on examples/basic-sc/README.md:
````diff
@@ -69,17 +69,6 @@ This means that `cartographer` by itself is not very useful - its powers arise
-	-f https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.30.0/release.yaml
-```
-
 - [kpack], for providing an opinionated way of continuously building container
````
This indentation seems off (not a new problem). 

I think we should rip out the installation instructions for each component and just put to the install docs of that component (that way we don't need to maintain versions and we don't rely on kapp).

```
- [kpack](https://github.com/pivotal/kpack/blob/main/docs/install.md)
- [source-controller](this may need an example...)
```

emmjohnson commented on examples/basic-sc/README.md:
```diff
@@ -140,11 +129,20 @@ registry (see [kpack secrets] to know more about how kpack makes use of it).
-cartographer-specific files that an App Operator would submit, and
-[./developer](./developer), containing Kubernetes objects that a developer
-would submit (yes, just a [Workload] and a [Pipeline]!)
+This example uses two directories with sub-directories of kubernetes resources:
```
For line 122 - Feels like we are missing a prereq -  a container image registry. We've given so much detail and now we just say you are on your own

emmjohnson replied:
More on #4. I think it is just confusing. There is nothing to actually do here.

emmjohnson commented on examples/basic-sc/README.md:
```diff
@@ -140,11 +129,20 @@ registry (see [kpack secrets] to know more about how kpack makes use of it).
-[./developer](./developer), containing Kubernetes objects that a developer
-would submit (yes, just a [Workload] and a [Pipeline]!)
+This example uses two directories with sub-directories of kubernetes resources:
+[../shared](../shared) and [.](.).
```
this second link is really hard to see

emmjohnson commented on examples/basic-sc/README.md:
```diff
@@ -140,11 +129,20 @@ registry (see [kpack secrets] to know more about how kpack makes use of it).
+This example uses two directories with sub-directories of kubernetes resources:
+[../shared](../shared) and [.](.).
+
+The shared directory has a subdirectory for cluster-wide configuration: [./00-cluster](./00-cluster)
```
I don't see 00-cluster in shared. Only cluster

emmjohnson commented on examples/basic-sc/README.md:
````diff
@@ -180,18 +178,32 @@ registry:
 ```
 
+### Observing the example
+
````
We should make tree a prereq if we are going to use it

emmjohnson commented on examples/shared/developer/service-account.yaml:
```diff
@@ -45,16 +50,13 @@ rules:
       - runnables
-      - tasks
-      - taskruns
     verbs:
```
get is now a required verb

emmjohnson commented on examples/shared/app-operator/app-deploy.yaml:
```diff
@@ -11,125 +11,10 @@
-      source:
-        blob:
-          url: $(sources.source.url)$
-      build:
```
It is confusing that these are no longer called templates?

waciumawanjohi replied:
Added `template` to the file names.

emmjohnson commented on examples/basic-sc/README.md:
````diff
@@ -180,18 +178,32 @@ registry:
   password: a-very-hard-to-break-password
 ```
 
+Next the developer needs to give Cartographer permission to create the
````
Clearer examples of 168. For gcr, the image prefix needs to include gcr.io/project-name/registry/demo-

emmjohnson commented on examples/basic-sc/README.md:
````diff
@@ -308,7 +317,7 @@ With the goal of creating a software supply chain like mentioned above
 ```
 
-  source --> test --> image --> knative service
+  source --> image --> knative service
````
Is the intention here to actually apply the k8s yaml in this section? Or is it meant to just  show what all of the steps would be to construct?

emmjohnson replied:
Regarding line 308: I see 5 apps remaining - 

```
Name             Namespaces                             Lcs   Lca
dev-ctrl         default                                true  47m
gitops-toolkit   (cluster),gitops-toolkit               true  2h
kapp-controller  (cluster),kapp-controller,kube-system  true  2h
knative-serving  (cluster),knative-serving              true  2h
kpack            (cluster),kpack                        true  2h
```

waciumawanjohi replied:
> I see 5 apps remaining

Uninstall has been updated to point to the documentation of knative/kpack/fluxcd/kapp. I believe that resolves this concern. Let me know if I've missed your intention.

waciumawanjohi replied:
> Is the intention here to actually apply the k8s yaml in this section?

The section above walks user through appyling, observing and tearing down. This section is to explain what's happened and to highlight important details. I've changed the heading from `step-by-step` to `Explanation`

emmjohnson commented on examples/basic-sc/app-operator/supply-chain.yaml:
```diff
@@ -0,0 +1,53 @@
+
+  #
+  #     source-provider                   fluxcd/GitRepository
+  #        <--[src]-- image-builder     kpack/Image
```
columns slightly off

emmjohnson commented on examples/basic-sc/values.yaml:
```diff
@@ -14,9 +14,10 @@
-service_account_name: workload-service-account
-image_prefix: projectcartographer/demo-
+service_account_name: cartographer-example-basic-sc-sa
+image_prefix: a-project-id-i-own/an-image-prefix-i-like-
```
maybe we have examples for gcr.io, docker, harbor?

waciumawanjohi replied:
I added a note to `examples/basic-sc/README.md` showing an example of gcr. I'm not sure how we would have three live examples here. Do you mean examples that are commented out?

emmjohnson commented on examples/runnable-tekton/README.md:
````diff
@@ -0,0 +1,547 @@
+3. Install [tekton] (providing a mechanism to create pipelines for application testing, scanning, etc):
+
+```bash
+TEKTON_VERSION=0.30.0 kubectl apply -f https://storage.googleapis.com/tekton-releases/pipeline/previous/v$TEKTON_VERSION/release.yaml
````
this did not work as a one liner for me. worked on two lines

emmjohnson commented on examples/runnable-tekton/README.md:
```diff
@@ -0,0 +1,547 @@
+
+2. Install Cartographer. Refer to [README.md](../../README.md).
+
+3. Install [tekton] (providing a mechanism to create pipelines for application testing, scanning, etc):
```
I think tekton should always be capitalized

emmjohnson commented on examples/runnable-tekton/README.md:
````diff
@@ -0,0 +1,547 @@
+Now let's update the Runnable with a different SHA, one where the tests fail:
+
+```bash
+k patch runnable test --type merge --patch "$(cat 02-tests-fail/runnable-patch.yml)"
````
kubectl patch

emmjohnson commented on examples/runnable-tekton/README.md:
````diff
@@ -0,0 +1,547 @@
+Now let's update Runnable with a commit where tests again pass:
+
+```bash
+k patch runnable test --type merge --patch "$(cat 03-tests-pass/runnable-patch.yml)"
````
kubectl patch

emmjohnson commented on examples/runnable-tekton/README.md:
````diff
@@ -0,0 +1,547 @@
+k patch runnable test --type merge --patch "$(cat 02-tests-fail/runnable-patch.yml)"
+```
+
+We can see that Runnable has a new child tekton taskrun:
````
TaskRun (more in file)

emmjohnson commented on examples/runnable-tekton/README.md:
```diff
@@ -0,0 +1,547 @@
+
+## Step by step
+
+We want to test our source code in kubernetes. We're interested in doing so with objects that
```
Capitalize Kubernetes

emmjohnson commented on examples/runnable-tekton/README.md:
```diff
@@ -0,0 +1,547 @@
+
+#### Creating new immutable objects
+
+For simplicity sake, we're going to pretend that there is a k8s resource `SuccessJob`. This resource is just like
```
remove k8s shorthand

emmjohnson commented on examples/runnable-tekton/README.md:
```diff
@@ -0,0 +1,547 @@
+## Step by step
+
+We want to test our source code in kubernetes. We're interested in doing so with objects that
+are easy to update. But projects like [tekton] create immutable runs for each test. While tekton
```
I like how you break down Runnable!

emmjohnson commented on examples/testing-sc/README.md:
```diff
@@ -0,0 +1,278 @@
+
+### Configuring the example
+
+Follow the [example configuration steps in basic-sc](../basic-sc/README.md#configuring-the-example)
```
It might be worth pointing to configuring in basic-sc, but then having the ytt command here.

waciumawanjohi replied:
Addressed in the follow on story: https://github.com/vmware-tanzu/cartographer/pull/504

emmjohnson commented on examples/testing-sc/developer/serviceaccount.yaml:
```diff
@@ -0,0 +1,47 @@
+      - update
+      - delete
+      - patch
+      - watch
```
- get is now required

emmjohnson commented on examples/testing-sc/developer/serviceaccount.yaml:
```diff
@@ -0,0 +1,47 @@
+# Copyright 2021 VMware
```
maybe call it role or rolebinding instead of service account?

davidmirror-ops commented on examples/basic-sc/README.md:
````diff
@@ -50,107 +50,67 @@ objects would be set by the different personas in the system:
-  -f https://github.com/knative/serving/releases/download/v$KNATIVE_SERVING_VERSION/serving-core.yaml
-```
+This example uses two directories with sub-directories of Kubernetes resources:
+[../shared](../shared) and [this directory (.)](.).
````
`this directory` reads confusing, could it be specific to `basic-sc` as the actual folder name?

davidmirror-ops commented on examples/basic-sc/README.md:
```diff
@@ -50,107 +50,67 @@ objects would be set by the different personas in the system:
 
+Finally, in this and in the shared directory there is a subdirectory containing
+Kubernetes objects that a developer would submit: [./developer](./developer) and
+[../shared/developer](../shared/developer)
```
It's a great idea to organize the content into folders and subfolders but this section feels confusing; what's the purpose of the `shared` folder? (I could infer it but would be useful to make it more explicit). If it's content that is shared among all the examples, we could avoid `and in the shared directory` so it could be easier to read

waciumawanjohi replied:
You are correct, it is content that's shared among the examples. It's specified here because users will need to submit those objects. If we remove the mention of them, I'm concerned that users will skip that step and the example will not work.

waciumawanjohi replied:
If you have ideas on how to clarify that language, I'm happy to incorporate them! Should we just use the same language as the paragraph above:

```
There is a subdirectory of cartographer-specific files that an App Developer would submit
in both this and in the shared directory:
```

davidmirror-ops commented on examples/basic-sc/README.md:
```diff
@@ -50,107 +50,67 @@ objects would be set by the different personas in the system:
 to - because [kpack/Image] is going to be used for building container images
 based on the source code, we need to tell it where those images should go, and
-which credentials it should use.
+which credentials it should use (see [kpack secrets] to know more about how
```
I struggled a couple of times when running the example to actually understand that there is a requirement to create a secret to provide the necessary push permission. So probably making it a imperative statement: `Create a secret with the authentication details of the container registry you plan to use for the example see kpack docs (link) as a reference)`

davidmirror-ops commented on examples/basic-sc/README.md:
```diff
@@ -180,18 +140,32 @@ registry:
+Next the developer needs to give Cartographer permission to create the
+items that will be stamped out in the supply chain. That will be done
+with a service account, whose name will be referenced in the workload.
+You can update [`values.yaml`](./values.yaml) with the desired name:
```
Also it would be beneficial to make it a imperative statement : ` Update values.yaml with your registry details`

waciumawanjohi replied:
I updated this with imperative language from your suggestion, but then read the comment below noting that no change is needed to this value, so I've reverted it.

davidmirror-ops commented on examples/basic-sc/README.md:
```diff
@@ -180,18 +140,32 @@ registry:
+#@data/values
+---
+# a name that won't collide with other service accounts
+service_account_name: cartographer-example-basic-sc-sa
```
This statement made me go to `values.yaml` just to discover that the SA key value was already there. At this point, mentioning that the user would only need to go to the file if needs to modify the SA name, would be useful

waciumawanjohi replied:
As noted above, the language is meant to show that updating this is optional.

martyspiewak commented on examples/runnable-tekton/README.md:
```diff
@@ -0,0 +1,533 @@
+
+1. We want to read values from the TaskRun conditionally. Only if Succeeded is True do we want to pass the
+  url and revision values forward in the supply chain.
+2. Taskruns are immutable. We cannot update the TaskRun when the next commit SHA is created.
```
nitpick:
```suggestion
2. TaskRuns are immutable. We cannot update the TaskRun when the next commit SHA is created.
```

martyspiewak commented on examples/shared/app-operator/image-kpack-image.yaml:
```diff
@@ -0,0 +1,49 @@
+spec:
+  params:
+    - name: image_prefix
+      default: projectcartographer/demo-
```
Maybe it would be helpful if we changed this to something that is more clearly a placeholder, like you did in `values.yaml`?

martyspiewak reviewed (approved):
Lgtm!

waciumawanjohi committed:
Expand examples

Separate basic supply chain from supply chain with testing.
Provide example of runnable.
